README.md
##### disclaimer

This is some kind `hello world` project of custom views in `javascript` for me. 
So it may contain some not commonly used practices
among experts of `javascript` 

# Batman
Reusable circular slider class in javascript (as shown in image below)

![alt text](docs/images/circularslider.png)



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/) and
follows the convention of [Keep A Changelog](http://keepachangelog.com/).


## [Unreleased]
## [1.1.0] - 2021-05-08
### Changed
- Improving gradient on circular slider progress bar

## [1.0.0] - 2021-04-29
### Added
- Initial release


src/scripts/sandbox.js
window.onload = function () {
    init();
};


function init() {
    addSliderLegend(750, 10, 145, "#2d0050", "Transportation");//#61467b
    addSliderLegend(650, 10, 119, "#003a93", "Food");//#167fc3
    addSliderLegend(500, 10, 93, "#2a7800", "Insurance");//#55a328
    addSliderLegend(800, 20, 67, "#d96a01", "Entertainment")//##e78a37
    addSliderLegend(200, 25, 41, "#cb1002", "Health care")//##ee6054
}

function addSliderLegend(value, step, radius, color, label) {
    let sliderContainer = document.getElementById("slider-container");
    let legendContainer = document.getElementById("legend-container");
    let slider = new CircularSlider(sliderContainer, {radius: radius, color: color, value: value, maxValue: 1000, step: step});
    let legend = new CircularSliderLegend(legendContainer, {prefix: "$", color: color, label: label, value: value});
    slider.setCallback(legend.setValue);
}


src/scripts/circular_slider.js
class CircularSlider {
    container = undefined;
    canvas = undefined;
    color = "#F00FFF";
    minValue = 0;
    maxValue = 100;
    step = 10;
    radius = 170;
    settingMode = false;

    constructor(container) {
        this.container = container;
        this.insertCanvas();
        this.initListeners();
        this.draw();
    }

    initListeners() {
        this.canvas.addEventListener("mousemove", (event) => {
            let cursorX = event.clientX - this.container.offsetLeft;
            let cursorY = event.clientY - this.container.offsetTop;
            console.log(cursorX + "/" + cursorY);
        });
        this.canvas.addEventListener("mousedown", (event) => {
            console.log("mouse mousedown")
        });
        this.canvas.addEventListener("mouseup", (event) => {
            console.log("mouse mouseup")
        });
        this.canvas.addEventListener("mouseleave", (event) => {
            console.log("mouse mouseleave")
        });

    }


    insertCanvas() {
        this.canvas = document.createElement('canvas');
        this.canvas.id = "CursorLayer";
        this.canvas.style.width = '100%';
        this.canvas.style.height = '100%';
        this.canvas.width = this.container.offsetWidth;
        this.canvas.height = this.container.offsetHeight;
        this.canvas.style.zIndex = 8;
        this.canvas.style.position = "relative";
        this.canvas.style.border = "1px solid";
        this.container.appendChild(this.canvas)
    }

    draw() {
        console.log("draw method");
        let ctx = this.canvas.getContext("2d");
        ctx.beginPath();
        ctx.arc(this.canvas.width / 2, this.canvas.height / 2, this.radius, 0, 2 * Math.PI);
        ctx.strokeStyle = this.color;
        ctx.lineWidth = 25;
        ctx.stroke();
    }
}


src/scripts/circular-slider-legend.js
class CircularSliderLegend {
    container = undefined;
    wrapper = undefined;
    valueSpan = undefined;
    colorSpan = undefined;
    labelSpan = undefined;

    prefix = "";
    value = 0;
    label = "";
    color = "#00FFFF"

    constructor(container, options) {
        this.container = container;
        this.setValuesFromOptions(options);
        this.createViews();
        this.applyValuesOnSpans();
    }

    setValuesFromOptions(options) {
        this.prefix = "";
        this.label = "";
        if (options) {
            if (options.hasOwnProperty("prefix")) {
                this.prefix = options['prefix'];
            }
            if (options.hasOwnProperty("label")) {
                this.label = options['label'];
            }
            if (options.hasOwnProperty("color")) {
                this.color = options['color'];
            }
            if (options.hasOwnProperty("value")) {
                this.value = options['value'];
            }
        }
    }

    createViews() {
        this.wrapper = document.createElement('div');
        this.wrapper.classList.add("circular-slider-legend");
        this.container.appendChild(this.wrapper);
        this.valueSpan = document.createElement('span');
        this.valueSpan.classList.add("value-span");
        this.wrapper.appendChild(this.valueSpan);
        this.colorSpan = document.createElement('span');
        this.colorSpan.classList.add("color-span");
        this.wrapper.appendChild(this.colorSpan);
        this.labelSpan = document.createElement('span');
        this.labelSpan.classList.add("label-spans");
        this.wrapper.appendChild(this.labelSpan);
    }

    applyValuesOnSpans() {
        this.applyValueOnSpan();
        this.applyColorOnSpan();
        this.applyLabelOnSpan();
    }

    setValue = (value) => {
        this.value = value;
        this.applyValuesOnSpans();
    }

    applyValueOnSpan() {
        this.valueSpan.textContent = this.prefix + this.value
    }

    applyColorOnSpan() {
          this.colorSpan.style.background = `linear-gradient(to bottom, ${shadeColor(this.color,300)} ,  ${this.color})`;
    }

    applyLabelOnSpan() {
        this.labelSpan.textContent = this.label;
    }

}


src/scripts/circular-slider.js
const HALF_PI = Math.PI * 0.5;
const TWO_PI = Math.PI * 2;
const SVG_NS = "http://www.w3.org/2000/svg";

class Point {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }

    toString() {
        return "x:" + this.x + "y:" + this.y;
    }
}

class EventsListener {
    constructor(id, centerPoint, innerRadius, outerRadius, callback) {
        this.id = id;
        this.centerPoint = centerPoint;
        this.innerRadius = innerRadius;
        this.outerRadius = outerRadius;
        this.callback = callback;
    }
}

class EventsHandler {
    CURSOR_IDENTIFIER = "cursor";
    idToEventListener = {}
    centerPointToListOfEventListeners = {}
    activeEventListeners = {};
    resizeObserver = undefined;
    onResizeCallbacks = [];

    constructor() {
        this.init();
    }

    registerEventListener(id, centerPoint, innerRadius, outerRadius, pointerTouchEventCallback, resizeEventCallback) {
        let eventListener = new EventsListener(id, centerPoint, innerRadius, outerRadius, pointerTouchEventCallback);
        this.onResizeCallbacks.push(resizeEventCallback);
        this.idToEventListener[eventListener.id] = eventListener;
        this.insertEventListenerIntoCenterPointDict(eventListener);
    }

    insertEventListenerIntoCenterPointDict(eventListener) {
        if (!this.centerPointToListOfEventListeners.hasOwnProperty(eventListener.centerPoint)) {
            this.centerPointToListOfEventListeners[eventListener.centerPoint] = []
        }
        this.centerPointToListOfEventListeners[eventListener.centerPoint].push(eventListener);
    }

    removeEventListenerFromEventListenerIntoCenterPointDict(eventListener) {
        let eventListeners = this.centerPointToListOfEventListeners[eventListener.centerPoint];
        let index = -1;
        for (let i = 0; i < eventListeners.length; i++) {
            if (eventListeners[i].id === eventListener.id) {
                index = i;
                break;
            }
        }
        if (index !== -1) {
            eventListeners.splice(index, 1);
        }
    }

    updateRadiusOfEventListener(id, innerRadius, outerRadius) {
        this.idToEventListener[id].innerRadius = innerRadius;
        this.idToEventListener[id].outerRadius = outerRadius;
    }

    updateCenterPointOfEventListener(id, newCenterPoint) {
        let eventListener = this.idToEventListener[id];
        this.removeEventListenerFromEventListenerIntoCenterPointDict(eventListener);
        this.idToEventListener[id].centerPoint = newCenterPoint;
        this.insertEventListenerIntoCenterPointDict(eventListener);
    }


    init() {
        document.addEventListener("DOMContentLoaded", (event) => {
            this.resizeObserver = new ResizeObserver(entries => {
                for (let i = 0; i < this.onResizeCallbacks.length; i++) {
                    this.onResizeCallbacks[i]();
                }
            });
            this.resizeObserver.observe(document.body);
        });

        document.addEventListener("mousedown", (event) => {
            this.handleStartTouchCursorEvent(event, this.CURSOR_IDENTIFIER, event.clientX, event.clientY);
        });
        document.addEventListener("mousemove", (event) => {
            this.handleMoveTouchCursorEvent(event, this.CURSOR_IDENTIFIER, event.clientX, event.clientY);
        });
        document.addEventListener("mouseleave", (event) => {
            this.handleEndTouchCursorEvent(event, this.CURSOR_IDENTIFIER);
        });
        document.addEventListener("mouseup", (event) => {
            this.handleEndTouchCursorEvent(event, this.CURSOR_IDENTIFIER);
        });
        document.addEventListener("touchstart", (event) => {
            for (let i = 0; i < event.touches.length; i++) {
                this.handleStartTouchCursorEvent(event, event.touches[i].identifier, event.touches[i].clientX, event.touches[i].clientY);
            }
        }, {passive: false});
        document.addEventListener("touchmove", (event) => {
            for (let i = 0; i < event.touches.length; i++) {
                this.handleMoveTouchCursorEvent(event, event.touches[i].identifier, event.touches[i].clientX, event.touches[i].clientY);
            }
        }, {passive: false});
        document.addEventListener("touchcancel", (event) => {
            for (let i = 0; i < event.changedTouches.length; i++) {
                this.handleEndTouchCursorEvent(event, event.changedTouches[i].identifier);
            }
        }, {passive: false});
        document.addEventListener("touchend", (event) => {
            for (let i = 0; i < event.changedTouches.length; i++) {
                this.handleEndTouchCursorEvent(event, event.changedTouches[i].identifier);
            }
        }, {passive: false});
    }

    handleStartTouchCursorEvent(event, identifier, clientX, clientY) {
        for (let centerPoint in this.centerPointToListOfEventListeners) {
            let eventListeners = this.centerPointToListOfEventListeners[centerPoint];
            for (let i = 0; i < eventListeners.length; i++) {
                let eventListener = eventListeners[i];
                let distance = Math.sqrt(Math.pow((clientX - eventListener.centerPoint.x), 2) + Math.pow((clientY - eventListener.centerPoint.y), 2));
                if (distance >= eventListener.innerRadius && distance <= eventListener.outerRadius) {
                    this.activeEventListeners[identifier] = eventListener;
                    event.preventDefault();
                    this.processActiveTouchEvent(identifier, clientX, clientY);
                    break;
                }
            }
        }

    }

    handleMoveTouchCursorEvent(event, identifier, clientX, clientY) {
        if (this.activeEventListeners[identifier]) {
            event.preventDefault();
            this.processActiveTouchEvent(identifier, clientX, clientY);
        }
    }

    handleEndTouchCursorEvent(event, identifier) {
        if (this.activeEventListeners[identifier]) {
            event.preventDefault();
            delete this.activeEventListeners[identifier];
        }
    }

    processActiveTouchEvent(identifier, clientX, clientY) {
        let centerX = this.activeEventListeners[identifier].centerPoint.x;
        let centerY = this.activeEventListeners[identifier].centerPoint.y;
        let angle = Math.atan2(clientY - centerY, clientX - centerX);
        angle = angle + HALF_PI;
        if (angle < 0) {
            angle = angle + TWO_PI;
        }
        let anglePercentage = angle / TWO_PI;
        this.activeEventListeners[identifier].callback(anglePercentage);
    }
}


class CircularSlider {
    static nextId = 1;
    static eventHandler = new EventsHandler();

    static getRandomColor() {
        return "#" + Math.floor(Math.random() * 16777215).toString(16);
    }

    id = undefined;
    callback = undefined;
    container = undefined;
    circularSliderViewSvg = undefined;
    progressViewCircle1 = undefined;
    progressViewCircleOffset1 = 1;
    progressViewCircle2 = undefined;
    progressViewCircleOffset2 = 1;
    gridViewCircle = undefined;
    endDotViewCircle = undefined;


    activeProgressColorStart = CircularSlider.getRandomColor();
    activeProgressColorMiddle = CircularSlider.getRandomColor();
    activeProgressColorEnd = CircularSlider.getRandomColor();
    minValue = 0;
    value = 25;
    maxValue = 100;
    step = 50;
    radius = 170;


    inactiveProgressColor = "#cfd0d1";
    endDotRadius = 11;
    progressWidth = 18;
    lockRadius = false;

    offsetTop = 0;
    offsetLeft = 0;
    width = 0;
    height = 0;
    valueRangeSize = 0;
    steps = undefined;
    absoluteCenterPoint = undefined;
    relativeCenterPoint = undefined;
    innerRadius = undefined;
    outerRadius = undefined;
    circumference = undefined;


    // NOTE I make options optional, you do not need to provide them
    constructor(container, options) {
        this.id = CircularSlider.nextId++;
        this.container = container;


        this.setValuesFromOptions(options);
        this.calculateNotContainerRelatedStuff();
        this.createViews();
        this.increaseContainerSizeIfNeeded();
        this.calculateContainerRelatedStuff();
        this.positionMainView();
        this.drawProgressDot();


        CircularSlider.eventHandler.registerEventListener(
            this.id,
            this.absoluteCenterPoint,
            this.innerRadius,
            this.outerRadius,
            this.onPointerTouchEvent,
            this.onViewResized);
    }

    onPointerTouchEvent = (anglePercentage) => {
        let step = Math.ceil(anglePercentage * this.steps);
        this.value = step * (this.maxValue - this.minValue) / this.steps;
        if (this.callback) {
            this.callback(this.value);
        }
        this.drawProgressDot();
    }

    onViewResized = () => {
        this.calculateContainerRelatedStuff();
        this.positionMainView();
        this.updateEventListener();
    }


    setCallback(callback) {
        this.callback = callback;
    }


    positionMainView() {
        this.circularSliderViewSvg.style.top = `${this.offsetTop}`;
        this.circularSliderViewSvg.style.left = `${this.offsetLeft}`;
    }

    updateEventListener() {
        CircularSlider.eventHandler.updateCenterPointOfEventListener(this.id, this.absoluteCenterPoint);
    }


    drawProgressDot() {
        this.drawProgressSvg();
        this.positionDotSvgCircle();
    }

    setValuesFromOptions(options) {
        this.lockRadius = false;
        if (options) {
            if (options['radius']) {
                this.radius = options['radius'];
                this.lockRadius = true;
            }
            if (options['color']) {
                this.activeProgressColorStart = options['color'];
            }
            if (options['minValue']) {
                this.minValue = options['minValue'];
            }
            if (options['maxValue']) {
                this.maxValue = options['maxValue'];
            }
            if (options['value']) {
                this.value = options['value'];
            }
            if (options['step']) {
                this.step = options['step'];
            }
        }
    }

    calculateNotContainerRelatedStuff() {
        this.valueRangeSize = Math.abs(this.maxValue - this.minValue);
        this.steps = this.valueRangeSize / this.step;
        this.activeProgressColorMiddle = shadeColor(this.activeProgressColorStart, 150);
        this.activeProgressColorEnd = shadeColor(this.activeProgressColorStart, 300);
        this.circumference = this.radius * TWO_PI;
        this.innerRadius = this.radius - this.progressWidth / 2;
        this.outerRadius = this.radius + this.progressWidth / 2;


        this.height = this.width = (this.innerRadius + Math.max(this.endDotRadius * 2, this.progressWidth)) * 2;
        this.relativeCenterPoint = new Point(
            this.width / 2,
            this.height / 2
        );
    }

    increaseContainerSizeIfNeeded() {
        let containerBoundingRect = this.container.getBoundingClientRect();
        if (containerBoundingRect.width < this.width) {
            this.container.style.width = this.width + "px";
        }
        if (containerBoundingRect.height < this.height) {
            this.container.style.height = this.height + "px";
        }
    }

    calculateContainerRelatedStuff() {
        const containerBoundingRect = this.container.getBoundingClientRect();
        this.offsetTop = (containerBoundingRect.height - this.height) / 2;
        this.offsetLeft = (containerBoundingRect.width - this.width) / 2;

        this.absoluteCenterPoint = new Point(
            this.relativeCenterPoint.x + containerBoundingRect.x + this.offsetLeft,
            this.relativeCenterPoint.y + containerBoundingRect.y + this.offsetTop
        );
    }


    createViews() {
        this.circularSliderViewSvg = document.createElementNS(SVG_NS, 'svg');
        this.circularSliderViewSvg.setAttribute("width", this.width);
        this.circularSliderViewSvg.setAttribute("height", this.height);
        this.circularSliderViewSvg.classList.add("circular-slider");
        this.container.appendChild(this.circularSliderViewSvg);

        let defsView = document.createElementNS(SVG_NS, 'defs');
        let mask = document.createElementNS(SVG_NS, 'mask');
        mask.setAttribute("id", "mask" + this.id);
        mask.setAttribute("maskUnits", "userSpaceOnUse");

        let rect = document.createElementNS(SVG_NS, 'rect');
        rect.setAttribute("x", "0");
        rect.setAttribute("y", "0");
        rect.setAttribute("width", this.width);
        rect.setAttribute("height", this.height);
        mask.appendChild(rect);

        for (let i = 0; i < this.steps; i++) {
            let angle = ((i / this.steps) * TWO_PI) - HALF_PI;
            let startX = this.innerRadius * Math.cos(angle) + this.relativeCenterPoint.x;
            let startY = this.innerRadius * Math.sin(angle) + this.relativeCenterPoint.y;
            let endX = this.outerRadius * Math.cos(angle) + this.relativeCenterPoint.x;
            let endY = this.outerRadius * Math.sin(angle) + this.relativeCenterPoint.y;
            let lineView = document.createElementNS(SVG_NS, 'line');
            lineView.setAttribute("x1", startX);
            lineView.setAttribute("y1", startY);
            lineView.setAttribute("x2", endX);
            lineView.setAttribute("y2", endY);
            mask.appendChild(lineView);

        }
        defsView.appendChild(mask);
        this.circularSliderViewSvg.appendChild(defsView);

        this.gridViewCircle = document.createElementNS(SVG_NS, 'circle');
        this.gridViewCircle.classList.add('grid-circle');
        this.gridViewCircle.setAttribute("r", `${this.radius}`);
        this.gridViewCircle.setAttribute("stroke-width", this.progressWidth);
        this.gridViewCircle.setAttribute("cx", this.relativeCenterPoint.x);
        this.gridViewCircle.setAttribute("cy", this.relativeCenterPoint.y);
        this.gridViewCircle.setAttribute("stroke", this.inactiveProgressColor);
        this.gridViewCircle.setAttribute("mask", `url(#mask${this.id})`);
        this.circularSliderViewSvg.appendChild(this.gridViewCircle);


        // TODO make something more similar to the conic gradient, probably with path instead of circles :/
        let half = Math.round(this.width / 2);
        this.progressViewCircle1 = this.createProgressCircle(1, half, this.width, this.activeProgressColorStart, this.activeProgressColorMiddle);
        this.progressViewCircle2 = this.createProgressCircle(2, 0, half, this.activeProgressColorEnd, this.activeProgressColorMiddle);

        this.endDotViewCircle = document.createElementNS(SVG_NS, 'circle');
        this.endDotViewCircle.classList.add('end-dot-circle');
        this.endDotViewCircle.setAttribute("r", this.endDotRadius);
        this.circularSliderViewSvg.appendChild(this.endDotViewCircle);
    }

    createProgressCircle(count, startX, endX, startColor, endColor,) {

        let defsView = document.createElementNS(SVG_NS, 'defs');
        let mask = document.createElementNS(SVG_NS, 'mask');
        mask.setAttribute("id", `progress-mask${count}-${this.id}`);
        mask.setAttribute("maskUnits", "userSpaceOnUse");

        let rect = document.createElementNS(SVG_NS, 'rect');
        rect.setAttribute("x", "0");
        rect.setAttribute("y", startX);
        rect.setAttribute("width", this.height);
        rect.setAttribute("height", endX);
        mask.appendChild(rect);
        defsView.appendChild(mask);
        this.circularSliderViewSvg.appendChild(defsView);


        let linearGradient = document.createElementNS(SVG_NS, 'linearGradient');
        linearGradient.setAttribute("id", `gradient${count}-${this.id}`);
        linearGradient.setAttribute("x1", "0%");
        linearGradient.setAttribute("y1", "0%");
        linearGradient.setAttribute("x2", "100%");
        linearGradient.setAttribute("y2", "0%");

        let linearGradientStop = document.createElementNS(SVG_NS, 'stop');
        linearGradientStop.setAttribute("offset", "0%");
        linearGradientStop.setAttribute("stop-color", endColor);
        linearGradient.appendChild(linearGradientStop);

        linearGradientStop = document.createElementNS(SVG_NS, 'stop');
        linearGradientStop.setAttribute("offset", "100%");
        linearGradientStop.setAttribute("stop-color", startColor);
        linearGradient.appendChild(linearGradientStop);
        this.circularSliderViewSvg.appendChild(linearGradient);

        let progressViewCircle = document.createElementNS(SVG_NS, 'circle');
        progressViewCircle.classList.add('progress-circle');
        progressViewCircle.setAttribute("r", `${this.radius}`);
        progressViewCircle.setAttribute("stroke-width", this.progressWidth);
        progressViewCircle.setAttribute("cx", this.relativeCenterPoint.x);
        progressViewCircle.setAttribute("cy", this.relativeCenterPoint.y);
        progressViewCircle.setAttribute("stroke", `url(#gradient${count}-${this.id})`);
        progressViewCircle.setAttribute("mask", `url(#progress-mask${count}-${this.id})`);
        progressViewCircle.style.strokeDasharray = `${this.circumference} ${this.circumference}`;
        this.circularSliderViewSvg.appendChild(progressViewCircle);

        return progressViewCircle;
    }

    getValuePartOfRange() {
        return (Math.abs(this.value - this.minValue) / (this.valueRangeSize));
    }

    drawProgressSvg() {
        const valuePartOfRange = this.getValuePartOfRange();
        const offset = this.circumference - valuePartOfRange * this.circumference;
        if (valuePartOfRange <= 0.5 || (valuePartOfRange > 0.5 && this.progressViewCircleOffset1 < 0.5)) {
            this.progressViewCircleOffset1 = valuePartOfRange;
            this.progressViewCircle1.style.strokeDashoffset = `${offset}`;
        }
        if (valuePartOfRange >= 0.5 || (valuePartOfRange < 0.5 && this.progressViewCircleOffset2 > 0.5)) {
            this.progressViewCircleOffset2 = valuePartOfRange;
            this.progressViewCircle2.style.strokeDashoffset = `${offset}`;
        }


    }

    positionDotSvgCircle() {
        let middleAngle = this.getValuePartOfRange() * TWO_PI - HALF_PI
        let x = this.radius * Math.cos(middleAngle) + this.relativeCenterPoint.x;
        let y = this.radius * Math.sin(middleAngle) + this.relativeCenterPoint.y;
        this.endDotViewCircle.setAttribute("cx", x);
        this.endDotViewCircle.setAttribute("cy", y);
    }
}

